fix: optimize route indexing in `bench migrate`

hasnain2808 opened the pull request:
The final step in `bench migrate` is building search index for all routes.
For sites that contain lots of pages it takes more than an hour to complete this step
We build each page via Jinja and context which is not necessary for simple markdown pages
Added a flag to do that

hasnain2808 committed:
fix: add route building as a background job

hasnain2808 committed:
fix: be explicit with timeout and queue

gavindsouza reviewed (commented):
If this can take over an hour for big sites then running it as a background job may not be the best. Apart from arbitrarily set timeout values, this also means one worker is blocked executing this. This could delay other jobs.

Users will be able to use their sites if you just turn off maintenance mode before executing this step. Also, moving it to the very end (perhaps after the `finally` block) could be another option.

hasnain2808 committed:
feat: website_search_field in doctype doctype
website_search_field to declare the content field
remove the background job execution of search indexing

hasnain2808 committed:
fix: title  needed too

hasnain2808 committed:
feat: better indexing

hasnain2808 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into add-route-index-build-as-a-background-job

hasnain2808 committed:
fix: cache document page

hasnain2808 committed:
feat: option to not preload assets

hasnain2808 committed:
perf: let's preload assets

leela commented on frappe/search/website_search.py:
```diff
@@ -85,16 +87,23 @@ def parse_result(self, result):
+					content = frappe.utils.md_to_html(getattr(doc, doctype.website_search_field))
+					soup = BeautifulSoup(content, "html.parser")
+					text_content = soup.text if soup else ""
+					_items_to_index += [frappe._dict(title=doc.title, content=text_content, path=doc.route)]
```
Size of this depends on number of pages exists in the system. We can use generator expression here to reduce high memory usage.

leela commented:
> If this can take over an hour for big sites then running it as a background job may not be the best. Apart from arbitrarily set timeout values, this also means one worker is blocked executing this. This could delay other jobs.
> 
> Users will be able to use their sites if you just turn off maintenance mode before executing this step. Also, moving it to the very end (perhaps after the `finally` block) could be another option.

leela commented:
> The final step in `bench migrate` is building search index for all routes.
> For sites that contain lots of pages it takes more than an hour to complete this step
> It is not necessary to stop users from using the site when this step is running.
> So added this to a background job so that the users may keep using the site while the routes are getting updated

Taking one hour is a too long. We should optimize code base if we can. Instead of creating one background job, we can go for multiple jobs each handling set of pages.

hasnain2808 commented:
> > The final step in `bench migrate` is building search index for all routes.
> > For sites that contain lots of pages it takes more than an hour to complete this step
> > It is not necessary to stop users from using the site when this step is running.
> > So added this to a background job so that the users may keep using the site while the routes are getting updated
> 
> Taking one hour is a too long. We should optimize code base if we can. Instead of creating one background job, we can go for multiple jobs each handling set of pages.


Hey,
The description was older.
Fixed the actual issue
There's no need to enqueue
Now it takes realistic times
Making it draft will reopen

hasnain2808 committed:
fix: add column entries in doctypes
